woocommerce/woocommerce-android
# Handle order Add-ons loading failure

Summary
==========
Fixes issue #4696 by refactoring the `OrderedAddonViewModel` to get cached data even when fetching recent information doesn't succeed. Also, it introduces a series of improvements for failure state at the view meanwhile refactoring the responsibilities between Fragment and ViewModel

Screenshots
==========
| Data loading successful  | Data loading failed |
| ------------- | ------------- |
| ![Screenshot_20210908_200706](https://user-images.githubusercontent.com/5920403/132597151-0ddf9f49-5182-423d-9fac-3d6441a865bd.png) | ![Screenshot_20210908_195401](https://user-images.githubusercontent.com/5920403/132597156-e25637e6-9fcb-49d6-b0d3-72de7ed9c80a.png) |

How to Test
==========
Prerequisites: Have your site with the add ons plugin installed, configured, and with at least one product with add-ons and at least one order of the configured product including Global Add-ons.

⚠️ Note that you should configure both Product-specific and Global add-ons in your store in order to test this improvement properly.

## Data loading successful
1 - Launch the Woo app.
2 - Navigate to an order with a Product configured with Add-ons, verify if the View Add-ons Button is showing up for products that contain Add-ons inside the current Order, but doesn't show up for Products without Add-ons for that order or Products that don't have Add-ons configured at all.
3 - Click on View Add-ons and verify if the Ordered Addons view is displayed with all the expected data, including Global Add-ons data.

## Data loading failed

We shouldn't be able to reproduce this failure easily since the `View Add-ons` button shouldn't be visible if there's no Ordered Add-ons to display, so I suggest to try to reproduce this issue by hard-coding the failure state, to do so, go to `OrderedAddonsViewModel` and change the start method to look like this:
``` suggestion
    fun start(
        orderID: Long,
        orderItemID: Long,
        productID: Long
    ) = viewState.copy(isSkeletonShown = true).let { viewState = it }.also {
        launch(dispatchers.computation) {
            addonsRepository.updateGlobalAddonsSuccessfully()
            loadOrderAddonsData(orderID, orderItemID, productID)
                .let { null }
                ?.let { dispatchResult(it) }
                ?: handleFailure()
        }
    }
```
After doing this change, do the following:
1 - Launch the Woo app.
2 - Navigate to an order with a Product configured with Add-ons, verify if the View Add-ons Button is showing up for products that contain Add-ons inside the current Order, but doesn't show up for Products without Add-ons for that order or Products that don't have Add-ons configured at all.
3 - Click on View Add-ons and verify if the Ordered Addons view is displaying the failure dialog, and pressing OK dismiss the entire view back to Order Details.


Update release notes:
